Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,687 @@
+                                                            tcppart, 20, IPPROTO_TCP);
+
+      WriteToQueue(&m_in_frame[0], 0x36);
+    }
+
+    // clear any ack data
+    if (tcpdata->properties & TCP_FLAG_ACK)
+    {
+      for (auto& tcp_buf : network_ref[i].tcp_buffers)
+      {
+        if (tcp_buf.used && tcp_buf.seq_id < Common::swap32(tcpdata->acknowledgement_number))
+        {
+          tcp_buf.used = false;  // confirmed data received
+          if (!network_ref[i].ready && !network_ref[i].tcp_buffers[0].used)
+            network_ref[i].ready = true;
+        }
+      }
+    }
+  }
+}
+
+/// <summary>
+/// This is a litle hack, Mario Kart open some UDP port
+/// and listen to it. We open it on our side manualy.
+/// </summary>
+void CEXIETHERNET::BuiltInBBAInterface::InitMKDD()
+{
+  int i = GetAvaibleSlot(htons(26512));
+  if (network_ref[i].ip != 0)
+    return;
+  network_ref[i].ip = 0x08080808;  // change for ip
+  network_ref[i].local = htons(26512);
+  network_ref[i].remote = htons(26512);
+  network_ref[i].type = 17;
+  memcpy(&network_ref[i].bba_mac, &m_eth_ref->mBbaMem[BBA_NAFR_PAR0], Common::MAC_ADDRESS_SIZE);
+  network_ref[i].my_mac = fake_mac;
+  network_ref[i].from.sin_addr.s_addr = 0;
+  network_ref[i].from.sin_port = htons(26512);
+  network_ref[i].to.sin_addr.s_addr = 0xc701a8c0;
+  network_ref[i].to.sin_port = htons(26512);
+  network_ref[i].udp_socket.setBlocking(false);
+  if (network_ref[i].udp_socket.bind(26512) != sf::Socket::Done)
+  {
+    ERROR_LOG_FMT(SP1, "Couldn't open UDP socket");
+    return;
+  }
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::HandleUDPFrame(Common::EthernetHeader* hwdata,
+                                                       Common::IPv4Header* ipdata,
+                                                       Common::UDPHeader* udpdata, u8* data)
+{
+  sf::IpAddress target;
+
+  if (*(u32*)ipdata->destination_addr == 0)
+    *(u32*)ipdata->destination_addr = 0x0101a8c0;
+  // dns request
+  int i = GetAvaibleSlot(udpdata->source_port);
+  if (network_ref[i].ip == 0)
+  {
+    network_ref[i].ip = *(u32*)ipdata->destination_addr;  // change for ip
+    network_ref[i].local = udpdata->source_port;
+    network_ref[i].remote = udpdata->destination_port;
+    network_ref[i].type = 17;
+    memcpy(&network_ref[i].bba_mac, &m_eth_ref->mBbaMem[BBA_NAFR_PAR0], Common::MAC_ADDRESS_SIZE);
+    network_ref[i].my_mac = fake_mac;
+    network_ref[i].from.sin_addr.s_addr = *(u32*)&ipdata->destination_addr;
+    network_ref[i].from.sin_port = udpdata->destination_port;
+    network_ref[i].to.sin_addr.s_addr = *(u32*)&ipdata->source_addr;
+    network_ref[i].to.sin_port = udpdata->source_port;
+    network_ref[i].udp_socket.setBlocking(false);
+    if (network_ref[i].udp_socket.bind(htons(udpdata->source_port)) != sf::Socket::Done)
+    {
+      if (network_ref[i].udp_socket.bind(sf::Socket::AnyPort) != sf::Socket::Done)
```

Comment:
im not a c++ expert but wont this try to bind 2 socket?

Reply:
No it shouldn't. It will try to bind using the first part of the condition and if it fails it will try the second one.

https://en.cppreference.com/w/cpp/language/operator_logical
> Builtin operators && and || perform short-circuit evaluation (do not evaluate the second operand if the result is known after evaluating the first)